.changeset/clean-pandas-act.md

```diff
@@ -0,0 +1,58 @@
+---
+'@nhost/nhost-js': minor
+---
+
+Use the `graphql-request` client at `nhost.graphql`
+
+`nhost.graphql` is now a fully-configured [graphql-request](https://github.com/prisma-labs/graphql-request) client.
+It adds the following features:
+
+- better Typescript integration
+- ability to use the client with the [GraphQL Code Generator](https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-graphql-request)
+- batch requests with `nhost.graphql.batchRequests()`
+
+### Breaking changes
+
+Before:
+
+```js
+const { data, error } = await nhost.graphql.request('{ myQuery { id } }')
+```
+
+After:
+
+```js
+try {
+  const data = await nhost.graphql.request('{ myQuery { id } }')
+} catch (error) {
+  console.log(error)
+}
+```
```

Comment:
I'm not sure we should move away from the previous `{ data, error}` approach because we're using it in other functions like `auth` and `storage`.

Is it possible to still use `{ data, error }` even with this package update?

Reply:
We can, but the trade-off is it won't be compatible with graphql-codegen or another tool meant to work with graphql-request.
Graphql-request also returns errors (plural) whereas our current signature is singular

Reply:
I want to come back to this, why can't the `{ data, error }` pattern be compatible with graphql-codegen? 

What other tools do we expect people to use graphql-request with?

The fact that the previous signature was singular doesn't concern me much. The `data` object here will be different anyways.

Reply:
| Graphql-request also returns errors (plural) whereas our current signature is singular

I don't think we should "tie" our package's API to that of the underlying GraphQL package's API. We want to expose the API we think is "right" while being able to change from `graphql-request` to `urql` in the future, if we want to.

Reply:
> I want to come back to this, why can't the `{ data, error }` pattern be compatible with graphql-codegen?
> 
> What other tools do we expect people to use graphql-request with where the `{ data, error }` pattern would be incompatible?
> 
> The fact that the previous signature was singular doesn't concern me much. The `data` object here will be different anyways.

it will be imcompatible with the `graphql-request` codegen:
```graphql
query myUsers {
  users {
    id
  }
}
```

```ts
import { NhostClient } from '@nhost/nhost-js'
import { getSdk } from 'generated.ts'

const nhost = new NhostClient()
const { myUsers } = getSdk(client)

const result = await myUsers() // <---- we don't have control over this one
```

Reply:
Let's use the new pattern for codegen described here: https://the-guild.dev/blog/typed-document-node

Similar to this syntax:

```js
import { request } from 'graphql-request'
import { graphql } from '../gql/gql'

const GET POSTS = graphql(`
  query GetPosts {
    posts {
      id
      title
    }
  }
`)

request('https://some-endpoint/graphql', GET_POSTS)
```

graphql-request ([and others](https://github.com/dotansimha/graphql-typed-document-node#built-in-support)) have support for this using TypedDocumentNode. 

- https://github.com/prisma-labs/graphql-request/blob/c17f82138e5c0e7960fd1fb2ba9ef439323d57dc/src/index.ts#L290

The equivalent syntax for our SDK would be:


```js
import { nhost } from '@/utils/nhost'
import { graphql } from '../gql/gql'

const GET POSTS = graphql(`
  query GetPosts {
    posts {
      id
      title
    }
  }
`)

const { data, error } = nhost.graphql.request(GET_POSTS);
```

What do you think?

Reply:
why not

Reply:
then the `errors` vs `error`: we would only pick the first one?

Reply:
Hmm, I think we should use `errors` like this: `{ data, errors }`. This way, we don't magically hide anything from developers.

`errors` (plural) is what the GraphQL server responds with, so it makes sense to use `errors` instead of `error`.